src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+    }
+}
+
+impl From<PcmRate> for u8 {
+    fn from(value: PcmRate) -> Self {
+        match value {
+            PcmRate::VIRTIO_SND_PCM_RATE_5512 => 0,
+            PcmRate::VIRTIO_SND_PCM_RATE_8000 => 1,
+            PcmRate::VIRTIO_SND_PCM_RATE_11025 => 2,
+            PcmRate::VIRTIO_SND_PCM_RATE_16000 => 3,
+            PcmRate::VIRTIO_SND_PCM_RATE_22050 => 4,
+            PcmRate::VIRTIO_SND_PCM_RATE_32000 => 5,
+            PcmRate::VIRTIO_SND_PCM_RATE_44100 => 6,
+            PcmRate::VIRTIO_SND_PCM_RATE_48000 => 7,
+            PcmRate::VIRTIO_SND_PCM_RATE_64000 => 8,
+            PcmRate::VIRTIO_SND_PCM_RATE_88200 => 9,
+            PcmRate::VIRTIO_SND_PCM_RATE_96000 => 10,
+            PcmRate::VIRTIO_SND_PCM_RATE_176400 => 11,
+            PcmRate::VIRTIO_SND_PCM_RATE_192000 => 12,
+            PcmRate::VIRTIO_SND_PCM_RATE_384000 => 13,
+            _ => 0,
+        }
+    }
+}
+
+impl From<SoundFeatures> for String {
+    fn from(sound_features: SoundFeatures) -> Self {
+        let mut features = vec![];
+        if sound_features.contains(SoundFeatures::VIRTIO_F_NOTIFY_ON_EMPTY) {
+            features.push("NOTIFY_ON_EMPTY");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_ANY_LAYOUT) {
+            features.push("ANY_LAYOUT");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_INDIRECT_DESC) {
+            features.push("RING_INDIRECT_DESC");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_EVENT_IDX) {
+            features.push("RING_EVENT_IDX");
+        }
+        if sound_features.contains(SoundFeatures::UNUSED) {
+            features.push("UNUSED");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_VERSION_1) {
+            features.push("VERSION_1");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_ACCESS_PLATFORM) {
+            features.push("ACCESS_PLATFORM");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_RING_PACKED) {
+            features.push("RING_PACKED");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_IN_ORDER) {
+            features.push("IN_ORDER");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_ORDER_PLATFORM) {
+            features.push("ORDER_PLATFORM");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_SR_IOV) {
+            features.push("SR_IOV");
+        }
+        if sound_features.contains(SoundFeatures::VIRTIO_F_NOTIFICATION_DATA) {
+            features.push("NOTIFICATION_DATA");
+        }
+        format!("[{}]", features.join(", "))
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct VirtIOSoundConfig {
+    jacks: ReadOnly<u32>,
+    streams: ReadOnly<u32>,
+    chmaps: ReadOnly<u32>,
+}
+
+/// In virtIO v1.2, this enum should be called "Code".
+///
+/// To avoid ambiguity in its meaning, I use the term "CommandCode" here.
+#[repr(u32)]
+#[derive(FromPrimitive, IntoPrimitive)]
```

Comment:
Derive `Copy` and `Clone` too, and maybe `Debug`, `Eq` and `PartialEq` in case they come in handy later for debugging or testing.

Reply:
In the `VirtIOSndJackInfo` struct, there are several bitflags represented as u32 (or u64) according to the virtio v1.2 specification. Directly implementing `Debug` for `VirtIOSndJackInfo` cannot print detailed information about these bitflags; it only prints their u32 value. Therefore, it is necessary to use `bitflags::parser::to_writer` to format the u32 bitflags for clearer output.